### ISSUE TITLE
[Bug] CLI override interacts poorly with `@_here_` usage in defaults list

### ISSUE THREAD

**Issue by Jasha10** [facebookresearch/hydra#1803]
# 🐛 Bug
When the defaults list contains `- group@_here_: choice`, overriding from the command line gives unexpected results.

## To reproduce
```yaml
# conf/base.yaml
defaults:
  - _self_
  - env@_here_: env1
a: 1
b: 2
```
```yaml
# conf/env/env1.yaml
c: 5
d: 6
```
```yaml
# conf/env/env2.yaml
c: 7
d: 8
```
```python
# my_app.py
import hydra
@hydra.main(config_path="conf", config_name="base")
def my_app(cfg):
    print(cfg)
my_app()
```

** Summary of current CLI override behavior: **

`python my_app.py`: works as expected
`python my_app.py env@_global_=env2`: produces error when no error is expected
`python my_app.py +env@_global_=env2`: the error message is not accurate
`python my_app.py env=env2`: does not error, an error would be expected
`python my_app.py +env=env2`: produces an error when no error is expected

** detailed descriptions below: **

#### `python my_app.py`
works as expected: produces output `{'a': 1, 'b': 2, 'c': 5, 'd': 6}`

#### `python my_app.py env@_global_=env2`
does not work as expected: 
```text
Could not override 'env@_global_'.
Did you mean to override env?
To append to your default list use +env@_global_=env2
```
The expected behavior is to produce `{'a': 1, 'b': 2, 'c': 7, 'd': 8}`.

#### `python my_app.py +env@_global_=env2`
produces this error: `Multiple values for env. To override a value use 'override env: env2'`
The error message should say something like `"Multiple values for env@_global_."`

#### `python my_app.py env=env2`:
produces `{'a': 1, 'b': 2, 'c': 7, 'd': 8}`.
The expected behavior would be to error.

#### `python my_app.py +env=env2`
Produces this error: `Multiple values for env. To override a value use 'override env: env2'`
The expected behavior is to produce `{'a': 1, 'b': 2, 'c': 5, 'd': 6, 'env': {'c': 7, 'd': 8}}`.

## Additional context
Based on a question here: https://stackoverflow.com/questions/68984309/overriding-the-package-specified-in-default-list-from-cli

Edit:
Some of the above behavior with `_global_` may be related to #1784.

**Comment by omry** [comment]
> python my_app.py env=env2: does not error, an error would be expected

This this produce the desired result?
This is what I would expect to work in this case. `_here_` is an alias to an empty package. and in this case the effective package is empty.

**Comment by omry** [comment]
Took a closer look.
> python my_app.py env=env2:
> produces {'a': 1, 'b': 2, 'c': 7, 'd': 8}.
> The expected behavior would be to error.

Reading the question, this is EXACTLY what the author wants to achieve.

I think this behavior is good and I would like to keep it.

**Comment by Jasha10** [comment]
Yes, `{'a': 1, 'b': 2, 'c': 7, 'd': 8}` is exactly what the author wants.

I still think that the API is inconsistent here.
Under normal circumstances:
1. if your defaults list has `- env@foo: env1`, the cli override would be `env@foo=env2`.
2. if your defaults list has `- env: env1`, the cli override would be `env=env2`.
3. if your defaults list has `- env@_global_: env1`, the cli override would be `env@_global_=env2`.

In all of these cases, if your defaults list has `- <GROUP>@<PACKAGE>: <OPTION>`, the corresponding CLI override is `<GROUP>@<PACKAGE>=<OPTION>`.

But: if your defaults list has `- env@_here_: env`, the pattern is broken.
Since `env@_here_=env2` does not work on the CLI.
Writing `env=env2` on the CLI is a key collision with case 2 above. I think that the CLI override `env=env2` should put the config `env2` into the default package `env`, not into the empty package (which is the current behavior).

The consistent way to achieve what the author wants should be either `env@_global_=env2` or `env@_here_=env2`, but neither works currently.

**Comment by omry** [comment]
`_here_` and `_global_` are not package names.
The logic is that if the final package (which is to say - the actual config node the config data will be embedded into) is different than the default package, you need to use the fully qualified package name to override it.
This by the way goes both for packages that are defined in the defaults list and for packages defined in package headers.

For example, requiring that configs with `# @package _global_` will always need to be overridden from the command line with `_global_` is a SERIOUS breaking change because it's a very common usage pattern.

You are suggesting to go in that direction for the sake of people that use `_here_` in config groups (`foo@_here_ : bar`) which is extremely uncommon (`_here_` example are always in the non overridable form `foo/bar@_here_` in the documentation).

You are correct that this is sometimes inconsistent. I will want to think very hard about introducing a breaking change here.

I will want to take a closer look at the other diff changing the handling of `_global_` because I uncomfortable with the direction here.

**Comment by Jasha10** [comment]
I see your point. I hadn't realized that it would be such a big breaking change.

> For example, requiring that configs with `# @package _global_` will always need to be overridden from the command line with `_global_` is a SERIOUS breaking change because it's a very common usage pattern.

**Comment by omry** [comment]
After spending some time on #1784 I am warming up to that fix as it looks like it's not interacting with the package header.
See my comments there.

**Comment by Jasha10** [comment]
> See my comments there.

You were right, this issue is fixed by #1791.
Here is the behavior after the merge:

python my_app.py: produces `{'a': 1, 'b': 2, 'c': 5, 'd': 6}` as expected
python my_app.py env@_global_=env2: produces `{'a': 1, 'b': 2, 'c': 7, 'd': 8}` as expected
python my_app.py +env@_global_=env2: produces error `Multiple values for env@_global_`, as expected
python my_app.py env=env2: produces error `Could not override 'env'`, as expected
python my_app.py +env=env2: produces `{'a': 1, 'b': 2, 'c': 5, 'd': 6, 'env': {'c': 7, 'd': 8}}` as expected

The bugfix is a breaking change since previously `python my_app.py env=env2` was necessary to produce the author's desired result, and now it will be `python my_app.py env@_global_=env2`.
